builder.go

```diff
@@ -175,25 +206,54 @@ func containsEntry(metRequires []string, entry BuildPlanEntry) bool {
 	return false
 }
 
-type processMap map[string]launch.Process
+// orderedProcesses is a mapping from process types to Processes, it will preserve ordering.
+// processList is the ordered list.
+// we keep typeToProcess map in order to delete elements when others are overriding them.
+type orderedProcesses struct {
+	typeToProcess map[string]*list.Element
+	processList   *list.List
+}
 
-func (m processMap) add(l []launch.Process) {
-	for _, proc := range l {
-		m[proc.Type] = proc
+func newProcessStruct() orderedProcesses {
+	return orderedProcesses{
+		typeToProcess: make(map[string]*list.Element),
+		processList:   list.New(),
 	}
 }
 
-func (m processMap) list() []launch.Process {
-	var keys []string
-	for key := range m {
-		keys = append(keys, key)
+func (m orderedProcesses) add(listToAdd []launch.Process) ([]string, error) {
+	result := []string{}
+	for _, proc := range listToAdd {
+		// when we replace a default process type with a non-default process type, add to result
+		if p, ok := m.typeToProcess[proc.Type]; ok {
+			cast, success := (p.Value).(launch.Process)
+			if !success {
+				return []string{}, fmt.Errorf("can't cast an element from the list to a process")
+			}
+			if cast.Default && !proc.Default {
+				result = append(result, proc.Type)
+			}
+			m.processList.Remove(p)
+		}
+		m.processList.PushBack(proc)
+		m.typeToProcess[proc.Type] = m.processList.Back()
```

Comment:
I have to confess that this code feels a bit overly complex for me, but maybe that's just a limit of my understanding. Could `processMap` somehow also hold on to the index of the buildpack in `range b.Group.Group`, and then when we have the complete map, we sort it by that index?

Reply:
This is just a suggestion, maybe the current implementation is the best way of doing it...

Reply:
mmmm...this is an interesting suggestion. The problem that I see is about overriding. I mean, there could be some numbers in the range that won't exist and then it won't be very easy to sort this map.
If you think I'm wrong and you know how to implement it, I would love to discuss it!

Reply:
If we add a field to metadata.toml (per https://github.com/buildpacks/lifecycle/pull/520#discussion_r569489888) this code should go away... the builder could just keep track of a single "default process" and update it for each buildpack when processing the build result.

Reply:
Please see https://github.com/buildpacks/lifecycle/pull/520/commits/48b609971e977b9ae0976f961226955d0e4dbf3f.